internals/search.py

```diff
@@ -311,6 +312,73 @@ def _sort_by_total_order(
   return sorted_id_list
 
 
+def make_cache_key(
+  user_query: str, sort_spec: str | None, show_unlisted: bool,
+  show_deleted: bool, show_enterprise: bool, start: int, num: int,
+  name_only: bool) -> str:
+  """Return a redis key string to store cached search results."""
+  return '|'.join([
+      FeatureEntry.SEARCH_CACHE_KEY,
+      user_query, str(sort_spec), str(show_unlisted),
+      str(show_deleted), str(show_enterprise), str(start), str(num),
+      str(name_only),
+  ])
+
+
+def is_cacheable(user_query: str, name_only: bool):
+  """Return True if this user query can be stored and viewed by other users."""
+  if not name_only:
+    logging.info('Search query not cached: could be large')
+    return False
+
+  if ':me' in user_query:
+    logging.info('Search query not cached: personalized')
+    return False
+
+  if ('is:recently-reviewed' in user_query or
+      'now' in user_query or
+      'current_stable' in user_query):
+    logging.info('Search query not cached: time-based')
+    return False
+
+  logging.info('Search query can be cached')
+  return True
+
+def process_query_using_cache(
+  user_query: str,
+  sort_spec: str | None = None,
+  show_unlisted=False,
+  show_deleted=False,
+  show_enterprise=False,
+  start=0,
+  num=DEFAULT_RESULTS_PER_PAGE,
+  context: Optional[QueryContext] = None,
+  name_only=False,
+) -> tuple[list[dict[str, Any]], int]:
+  """"""
+  cache_key = make_cache_key(
+      user_query, sort_spec, show_unlisted, show_deleted, show_enterprise,
+      start, num, name_only)
+  if is_cacheable(user_query, name_only):
+    logging.info('Checking cache at %r', cache_key)
+    cached_result = rediscache.get(cache_key)
+    if cached_result:
+      logging.info('Found cached search result')
+      return cached_result
+
+  logging.info('Computing search result')
+  computed_result = process_query(
+      user_query, sort_spec=sort_spec, show_unlisted=show_unlisted,
+      show_deleted=show_deleted, show_enterprise=show_enterprise,
+      start=start, num=num, context=context, name_only=name_only)
+
+  if is_cacheable(user_query, name_only):
+    logging.info('Storing search result in cache: %r', cache_key)
+    rediscache.set(cache_key, computed_result)
```

Comment:
Let's set a short expiry date here. Less chance for issues

Reply:
Done.  I set it to 1 hour.  That is probably about the longest that it will actually stay cached during a normal day anyway, because we clear the cache when people do edits.